data/libs/prelude.icicle

```diff
@@ -112,6 +112,52 @@ numflips' v t
           end
  ~> case s | (s0,i) -> i end.
 
+# Quantiles from sample, same as in R.
+quantile_on_sorted_buckets k q array
+ =  let len          = length array
+ ~> let actual       = (double k) / (q * double (len - 1))
+ ~> let actual_ceil  = ceil  actual
+ ~> let actual_floor = floor actual
+ ~> let floor_weight = double actual_ceil - actual
+ ~> let val_floor    = index array actual_floor
+ ~> let val_ceil     = index array actual_ceil
+ ~> double val_floor + double val_ceil * double (1 - floor_weight).
+
+#######################
+### Example queries ###
+#######################
+
+# Count the days that an entity has spent in different states.
+# e.g. days_in state time = [("NSW", 20), ("QLD", 1)]
+days_in x t
+ = fold ds
+    = ((None, 0), map_create)
+    : case ds
+       |  ((None,         nv),     m)
+       -> ((Some (x, t),  nv + 1), m)
+       |  ((Some (v, tv), nv),     m)
+       -> (  let tx = t
+          ~> let nx = days between tv tx
+          ~> let n  = nv + nx
+          ~> case v == x
+              | True
+              -> ((Some (x, tx), nv + nx), m)
+              | False
+              -> (  let m' = (  let nv_ = map_lookup v m
+                             ~> case nv_
+                                 | None
+                                 -> map_insert v n m
+                                 | Some nv'
+                                 -> map_insert v (nv' + n) m
+                                end
+                              )
+                 ~> (Some (x, tx), 0, m')
+                 )
+             end
+          )
+      end
+ ~> snd ds.
```

Comment:
this is amazing. would be good to have some test cases for these two, partly for sanity and partly as examples